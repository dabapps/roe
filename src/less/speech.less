.create-speech-styles (@background: @grey-lightest, @color: @grey-light, @position: left) {
  position: relative;
  margin: 0 10px;
  border-radius: 10px;
  border-bottom-@{position}-radius: 0;
  border: 1px solid @background;
  background-color: @color;
  // text-align: @position;
  color: @speech-text-color-default;

  &::before,
  &::after{
    content: '';
    position: absolute;
    width: 10px;
    height: 10px;
  }
}

.create-speech (@type, @speech-border-default: @grey-lightest, @speech-color-default: @grey-light, @position) {
  &.@{type} {
    .create-speech-styles(@speech-border-default, @speech-color-default, @position);

    &::before {
      left: inherit;
      right: inherit;
      @{position}: -11px;
      bottom: -1px;
      background: radial-gradient(circle at top @position, transparent 9px, @speech-border-default 10px);
    }

    &::after {
      left: inherit;
      right: inherit;
      @{position}: -9px;
      bottom: 0;
      background: radial-gradient(circle at top @position, transparent 9px, @speech-color-default 10px);
    }
  }
}

.speech {
  & .bubble {
    padding: 10px;
  }
  .create-speech-styles();
  .create-speech(received, @speech-border-default-received, @speech-background-default-received, left);
  .create-speech(sent, @speech-border-default-sent, @speech-background-default-sent, right);
}
