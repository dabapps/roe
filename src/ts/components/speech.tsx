import * as classNames from 'classnames';
import * as React from 'react';
import { HTMLProps, PureComponent } from 'react';
import { ComponentProps } from '../types';

export interface SpeechProps extends ComponentProps, HTMLProps<HTMLElement> {
  /**
   *  Position the speech bubble tail on the right
   */
  sent?: boolean;
  /**
   * Position the speech bubble tail on the left
   */
  received?: boolean;
}

/**
 * Speech bubble component for displaying conversations / messages.
 */
export class Speech extends PureComponent<SpeechProps, {}> {
  public render() {
    const {
      className,
      children,
      sent,
      received,
      component: Component = 'div',
      ...remainingProps
    } = this.props;

    return (
      <Component
        {...remainingProps}
        className={classNames(
          'speech',
          sent && 'sent',
          received && 'received',
          className
        )}
      >
        <div className="bubble">{children}</div>
      </Component>
    );
  }
}

export default Speech;
