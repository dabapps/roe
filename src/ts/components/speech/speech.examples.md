#### Example

```js
<Row>
  <Column className="margin-bottom-base">
    <Speech sent>
      <p>Hello</p>
    </Speech>
  </Column>
  <Column>
    <Speech received>
      <p>Hi</p>
    </Speech>
  </Column>
</Row>
```

#### Less variables

```less
@speech-arrow-size: 10px;
@speech-border-radius: 10px;
@speech-text-color-default: @grey-dark;
@speech-border-default-received: @grey-light;
@speech-background-default-received: @grey-lightest;
@speech-border-default-sent: @grey-light;
@speech-background-default-sent: @grey-lightest;
```
