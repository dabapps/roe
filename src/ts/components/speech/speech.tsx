import * as classNames from 'classnames';
import * as React from 'react';
import { HTMLProps, PureComponent } from 'react';
import { ComponentProps } from '../../types';

export interface SpeechProps extends ComponentProps, HTMLProps<HTMLElement> {
/**
 *  Position the speech bubble arrow thingy on the right
 */
  sent ?: boolean;
/**
 * Position the speech bubble arrow thingy on the left
 */
  received ?: boolean;
}
/**
 * Speech Bubble Component
 */
export class Speech extends PureComponent<SpeechProps, {}> {
  public render() {
    const {
      className,
      children,
      sent,
      received,
      component: Component = 'div',
      ...remainingProps
    } = this.props;

    return (
      <Component
        {...remainingProps}
        className={classNames(
          'speech',
          sent ? 'sent' : null,
          received ? 'received' : null,
          className,
        )}
      >
        <div className="bubble">
          {children}
        </div>
      </Component>
    );
  }
}

export default Speech;
