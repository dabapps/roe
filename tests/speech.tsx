import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Speech } from '../src/ts/components/speech/speech';

describe('Speech', () => {
  it('should match snapshot', () => {
    const tree = renderer.create(
      <Speech sent>
        I love kittens!
      </Speech>
    );

    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot', () => {
    const tree = renderer.create(
      <Speech received>
        I love puggs!
      </Speech>
    );

    expect(tree).toMatchSnapshot();
  });

});
