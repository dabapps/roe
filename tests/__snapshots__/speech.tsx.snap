// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Speech should match snapshot 1`] = `
<div
  className="speech sent"
>
  <div
    className="bubble"
  >
    <p>
      I love kittens!
    </p>
  </div>
</div>
`;

exports[`Speech should match snapshot 2`] = `
<div
  className="speech received"
>
  <div
    className="bubble"
  >
    <p>
      I love puggs!
    </p>
  </div>
</div>
`;
